cloudsync/providers/file.py

```diff
@@ -0,0 +1,504 @@
+            raise ex.CloudFileExistsError
+
+    @property  # type: ignore
+    def latest_cursor(self):
+        return self._latest_cursor
+
+    @property  # type: ignore
+    def current_cursor(self):
+        return self._cursor
+
+    @current_cursor.setter  # type: ignore
+    def current_cursor(self, val):
+        if val is None:
+            val = self.latest_cursor
+        if not isinstance(val, int) and val is not None:
+            raise ex.CloudCursorError(val)
+        if val > (self.latest_cursor + 1):
+            raise ex.CloudCursorError("%s is not a valid cursor" % val)
+
+        self._cursor = val
+
+    def _convert_watchdog_event(self, event):
+        otype = OType.DIRECTORY if event.is_directory else OType.FILE
+        fpath = event.src_path
+        oid = self._fpath_to_oid(fpath)
+        exists = True
+        prior_oid = None
+
+        if hasattr(event, "dest_path"):
+            prior_oid = oid
+            fpath = event.dest_path
+            oid = self._fpath_to_oid(fpath)
+
+        if type(event) in (watchdog_events.DirDeletedEvent, watchdog_events.FileDeletedEvent):
+            exists = False
+
+        if type(event) in (watchdog_events.DirModifiedEvent, watchdog_events.FileModifiedEvent) and exists:
+            return None
+
+        try:
+            stat = os.stat(fpath)
+            mtime = stat.st_mtime
+        except FileNotFoundError:
+            exists = False
+            mtime = time.time()
+
+        ret = Event(otype=otype, hash=None, path=self._trim_ns(fpath), oid=oid, exists=exists, prior_oid=prior_oid, mtime=mtime)
+        return ret
+
+    def _on_any_event(self, event):
+        ev = self._convert_watchdog_event(event)
+        if not ev:
+            return
+        with self._evlock:
+            self._latest_cursor += 1
+            ev.new_cursor = self._latest_cursor
+            self._events.append(ev)
+            assert self._events[ev.new_cursor - 1] is ev
+            assert len(self._events) + self._evoffset == self._latest_cursor
+
+    def events(self) -> typing.Generator[Event, None, None]:
+        if self._cursor < self._evoffset:
+            self._cursor = self._evoffset
+
+        while self._cursor < self._latest_cursor:
+            pe = None
+            with self._evlock:
+                if self._cursor < self._latest_cursor:
+                    pe = self._events[self._cursor - self._evoffset]
+                    self._cursor += 1
+            if pe is not None:
+                yield pe
+        while len(self._events) > self._event_window:
+            with self._evlock:
+                self._events.popleft()
+                self._evoffset += 1
+        assert len(self._events) + self._evoffset == self._latest_cursor
+
+    @staticmethod
+    def _oid_to_fpath(oid):
+        return oid
+
+    def _fpath_to_oid(self, path):
+        return self.normalize_path(path)
+
+    def upload(self, oid, file_like, metadata=None) -> OInfo:
+        with self._api():
```

Comment:
`self._api()` returns a client, which I get is just `self`, but it's still good organization to use
```suggestion
        with self._api() as client:
```
and then use client everywhere inside the block instead of self. That way it is easier to evaluate if we are using `self._api()` in the necessary places, plus it is makes it easier to enhance the capabilities inside the `with` block later if we are using a special variable to access the "api"

Reply:
or maybe `with self._api() as client` should return os instead of self, and then we could use `client.path.join()` instead of `os.path.join()` kind of thing

Reply:
it's sort of meaningless.... i'm just using it to convert exceptions, not for any notion of there being a "client".   imo: it would be confusing to pretend client is or can be anything meaningful other than an exception wrapper